File: foodbridge-backend/src/main/java/com/foodbridge/entity/Donation.java
package com.foodbridge.entity;

import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.LastModifiedDate;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@Document(collection = "donations")
public class Donation {

    @Id
    private String id;

    @Indexed
    private String donorId;

    private String donorName;
    private String organizationName;
    private String foodName;
    private String foodType;
    private Integer quantity;
    private String unit;
    private String location;
    private Double latitude;
    private Double longitude;
    private Instant pickupTime;
    private Instant expiryTime;
    private String description;
    private StoredFile image;
    private String priority;
    private Long remainingPickupWindowMinutes;
    private Boolean manualReviewRequired;
    private String manualReviewReason;

    @Builder.Default
    private Map<String, Object> aiAnalysis = new HashMap<>();

    @Builder.Default
    private Map<String, Object> validation = new HashMap<>();

    @Builder.Default
    private Map<String, Object> locationDetails = new HashMap<>();

    @Builder.Default
    private List<Map<String, Object>> recommendedNgos = new ArrayList<>();

    @Indexed
    @Builder.Default
    private String status = "available";

    private String assignedReceiverId;
    private String assignedReceiverName;
    private String assignedVolunteerId;
    private String assignedVolunteerName;

    @CreatedDate
    private Instant createdAt;

    @LastModifiedDate
    private Instant updatedAt;
}

File: foodbridge-backend/src/models/Donation.js
import mongoose from "mongoose";

const donationSchema = new mongoose.Schema(
  {
    donorId: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true, index: true },
    donorName: { type: String, required: true, trim: true },
    organizationName: { type: String, required: true, trim: true },
    foodName: { type: String, required: true, trim: true },
    foodType: { type: String, enum: ["Veg", "Non-Veg", "Both"], required: true },
    quantity: { type: Number, required: true, min: 1 },
    unit: { type: String, enum: ["Meals", "Plates", "Boxes", "Packs", "Kg"], default: "Meals" },
    location: { type: String, required: true, trim: true },
    pickupTime: { type: Date, required: true },
    expiryTime: { type: Date, required: true },
    description: { type: String, trim: true, maxlength: 500, default: "" },
    image: { type: mongoose.Schema.Types.Mixed },
    status: { type: String, enum: ["available", "requested", "assigned", "collected"], default: "available", index: true },
    assignedReceiverId: { type: mongoose.Schema.Types.ObjectId, ref: "User", default: null },
    assignedReceiverName: { type: String, default: "" },
  },
  { timestamps: true }
);

export default mongoose.model("Donation", donationSchema);
